File: frontend/src/api/bootstrap.ts
import { apiClient, type ApiResponse } from './client';

// --- Types ---

export interface BootstrapScores {
    architecture_score: number;
    contract_density: number;
    risk_score: number;
    alignment_score: number;
}

import type { components } from './generated/schema';

export type ImportSession = components['schemas']['ImportSession'];

export interface BootstrapConfidence {
    project_overview: number;
    tech_stack: number;
    modules: number;
    apis: number;
    data_models: number;
    validation_rules: number;
    contracts: number;
    current_state: number;
    risks: number;
    change_sensitivity: number;
}

export interface BootstrapModule {
    name: string;
    description: string;
    responsibilities?: string[];
    risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'UNKNOWN';
    change_sensitivity: 'LOW' | 'MEDIUM' | 'HIGH' | 'UNKNOWN';
}

export interface BootstrapApiEntry {
    endpoint: string;
    method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
    auth_type: string;
    request_schema?: Record<string, any>;
    response_schema?: Record<string, any>;
}

export interface BootstrapEntity {
    name: string;
    relationships?: Record<string, any>[];
    constraints?: Record<string, any>[];
}

export interface BootstrapContractEntry {
    name: string;
    contract_type: string;
    schema?: Record<string, any>;
    source_module: string;
    stability_score: number;
}

export interface ProjectIntelligenceSnapshot {
    id: string;
    project_id: string;
    version: number;
    snapshot_json: any;
    scores: BootstrapScores;
    confidence: BootstrapConfidence;
    modules?: BootstrapModule[];
    entities?: BootstrapEntity[];
    apis?: BootstrapApiEntry[];
    contracts?: BootstrapContractEntry[];
    created_at: string;
    // Flatten score fields for convenience
    architecture_score: number;
    contract_density: number;
    risk_score: number;
    alignment_score: number;
}

export interface BootstrapPromptResponse {
    prompt: string;
    project_name: string;
    project_id: string;
}

export interface BootstrapIngestResponse {
    snapshot: ProjectIntelligenceSnapshot;
    scores: BootstrapScores;
    confidence: BootstrapConfidence;
    warnings: string[];
}

export interface BootstrapDiffResult {
    from_snapshot_id: string | null;
    to_snapshot_id: string;
    from_version: number;
    to_version: number;
    added_modules?: BootstrapModule[];
    removed_modules?: BootstrapModule[];
    added_apis?: BootstrapApiEntry[];
    removed_apis?: BootstrapApiEntry[];
    added_entities?: BootstrapEntity[];
    removed_entities?: BootstrapEntity[];
    score_changes?: Record<string, number>;
}

export interface BootstrapPayload {
    project_overview?: Record<string, any>;
    tech_stack?: Record<string, any>;
    modules?: Record<string, any>[];
    apis?: Record<string, any>[];
    data_models?: Record<string, any>[];
    validation_rules?: Record<string, any>[];
    contracts?: Record<string, any>[];
    current_state?: Record<string, any>;
    risks?: Record<string, any>[];
    change_sensitivity?: Record<string, any>[];
}

// --- API Functions ---

export const bootstrapApi = {
    generatePrompt: async (projectId: string): Promise<BootstrapPromptResponse> => {
        const response = await apiClient.post<ApiResponse<BootstrapPromptResponse>>(
            `projects/${projectId}/bootstrap/generate-prompt`
        );
        return response.data.data;
    },

    ingestBootstrap: async (projectId: string, payload: BootstrapPayload): Promise<BootstrapIngestResponse> => {
        const response = await apiClient.post<ApiResponse<BootstrapIngestResponse>>(
            `projects/${projectId}/bootstrap/ingest`,
            payload
        );
        return response.data.data;
    },

    listSnapshots: async (projectId: string): Promise<ProjectIntelligenceSnapshot[]> => {
        const response = await apiClient.get<ApiResponse<ProjectIntelligenceSnapshot[]>>(
            `projects/${projectId}/bootstrap/snapshots`
        );
        return response.data.data;
    },

    getSnapshot: async (projectId: string, snapshotId: string): Promise<ProjectIntelligenceSnapshot> => {
        const response = await apiClient.get<ApiResponse<ProjectIntelligenceSnapshot>>(
            `projects/${projectId}/bootstrap/snapshots/${snapshotId}`
        );
        return response.data.data;
    },

    getLatestSnapshot: async (projectId: string): Promise<ProjectIntelligenceSnapshot> => {
        const response = await apiClient.get<ApiResponse<ProjectIntelligenceSnapshot>>(
            `projects/${projectId}/bootstrap/latest`
        );
        return response.data.data;
    },

    diffSnapshots: async (
        projectId: string,
        fromSnapshotId?: string,
        toSnapshotId?: string
    ): Promise<BootstrapDiffResult> => {
        const response = await apiClient.post<ApiResponse<BootstrapDiffResult>>(
            `projects/${projectId}/bootstrap/diff`,
            {
                from_snapshot_id: fromSnapshotId || undefined,
                to_snapshot_id: toSnapshotId || undefined,
            }
        );
        return response.data.data;
    },

    getLatestImportSession: async (projectId: string): Promise<ImportSession> => {
        const response = await apiClient.get<ApiResponse<ImportSession>>(
            `projects/${projectId}/bootstrap/session`
        );
        return response.data.data;
    },

    createProject: async (data: { name: string }): Promise<{ id: string; name: string }> => {
        const response = await apiClient.post<ApiResponse<{ id: string; name: string }>>(
            'projects',
            data
        );
        return response.data.data;
    },
};

File: backend/internal/api/bootstrap.go
package api

import (
	"database/sql"
	"net/http"

	"github.com/google/uuid"
	"github.com/labstack/echo/v4"
	"github.com/scott/specforge/internal/app"
	"github.com/scott/specforge/internal/domain"
)

type BootstrapHandler struct {
	service app.BootstrapService
}

func NewBootstrapHandler(service app.BootstrapService) *BootstrapHandler {
	return &BootstrapHandler{service: service}
}

func (h *BootstrapHandler) GeneratePrompt(c echo.Context) error {
	projectID, err := uuid.Parse(c.Param("projectId"))
	if err != nil {
		return ErrorResponse(c, http.StatusBadRequest, "INVALID_ID", "invalid project id", err.Error())
	}

	prompt, projectName, err := h.service.GeneratePrompt(c.Request().Context(), projectID)
	if err != nil {
		if err == sql.ErrNoRows {
			return ErrorResponse(c, http.StatusNotFound, "NOT_FOUND", "project not found", err.Error())
		}
		return ErrorResponse(c, http.StatusInternalServerError, "INTERNAL_ERROR", "failed to generate prompt", err.Error())
	}

	return SuccessResponse(c, http.StatusOK, map[string]interface{}{
		"prompt":       prompt,
		"project_name": projectName,
		"project_id":   projectID,
	})
}

func (h *BootstrapHandler) IngestBootstrap(c echo.Context) error {
	projectID, err := uuid.Parse(c.Param("projectId"))
	if err != nil {
		return ErrorResponse(c, http.StatusBadRequest, "INVALID_ID", "invalid project id", err.Error())
	}

	var payload domain.BootstrapPayload
	if err := c.Bind(&payload); err != nil {
		return ErrorResponse(c, http.StatusBadRequest, "INVALID_REQUEST", "failed to parse bootstrap payload", err.Error())
	}

	snapshot, warnings, err := h.service.IngestBootstrap(c.Request().Context(), projectID, payload)
	if err != nil {
		if err == sql.ErrNoRows {
			return ErrorResponse(c, http.StatusNotFound, "NOT_FOUND", "project not found", err.Error())
		}
		return ErrorResponse(c, http.StatusInternalServerError, "INTERNAL_ERROR", "failed to ingest bootstrap", err.Error())
	}

	return SuccessResponse(c, http.StatusCreated, map[string]interface{}{
		"snapshot": snapshot,
		"scores": map[string]interface{}{
			"architecture_score": snapshot.ArchitectureScore,
			"contract_density":   snapshot.ContractDensity,
			"risk_score":         snapshot.RiskScore,
			"alignment_score":    snapshot.AlignmentScore,
		},
		"confidence": snapshot.ConfidenceJSON,
		"warnings":   warnings,
	})
}

func (h *BootstrapHandler) ListSnapshots(c echo.Context) error {
	projectID, err := uuid.Parse(c.Param("projectId"))
	if err != nil {
		return ErrorResponse(c, http.StatusBadRequest, "INVALID_ID", "invalid project id", err.Error())
	}

	snapshots, err := h.service.ListSnapshots(c.Request().Context(), projectID)
	if err != nil {
		return ErrorResponse(c, http.StatusInternalServerError, "INTERNAL_ERROR", "failed to list snapshots", err.Error())
	}

	return SuccessResponse(c, http.StatusOK, snapshots)
}

func (h *BootstrapHandler) GetSnapshot(c echo.Context) error {
	snapshotID, err := uuid.Parse(c.Param("snapshotId"))
	if err != nil {
		return ErrorResponse(c, http.StatusBadRequest, "INVALID_ID", "invalid snapshot id", err.Error())
	}

	snapshot, err := h.service.GetSnapshot(c.Request().Context(), snapshotID)
	if err != nil {
		return ErrorResponse(c, http.StatusNotFound, "NOT_FOUND", "snapshot not found", err.Error())
	}

	return SuccessResponse(c, http.StatusOK, snapshot)
}

func (h *BootstrapHandler) GetLatestSnapshot(c echo.Context) error {
	projectID, err := uuid.Parse(c.Param("projectId"))
	if err != nil {
		return ErrorResponse(c, http.StatusBadRequest, "INVALID_ID", "invalid project id", err.Error())
	}

	snapshot, err := h.service.GetLatestSnapshot(c.Request().Context(), projectID)
	if err != nil {
		if err == sql.ErrNoRows {
			return ErrorResponse(c, http.StatusNotFound, "NOT_FOUND", "project not found", err.Error())
		}
		return ErrorResponse(c, http.StatusNotFound, "NOT_FOUND", "no snapshots found", err.Error())
	}

	return SuccessResponse(c, http.StatusOK, snapshot)
}

func (h *BootstrapHandler) GetLatestImportSession(c echo.Context) error {
	projectID, err := uuid.Parse(c.Param("projectId"))
	if err != nil {
		return ErrorResponse(c, http.StatusBadRequest, "INVALID_ID", "invalid project id", err.Error())
	}

	session, err := h.service.GetLatestImportSession(c.Request().Context(), projectID)
	if err != nil {
		return ErrorResponse(c, http.StatusInternalServerError, "INTERNAL_ERROR", "failed to fetch session", err.Error())
	}

	if session == nil {
		return ErrorResponse(c, http.StatusNotFound, "NOT_FOUND", "no import session found", "no import session found")
	}

	return SuccessResponse(c, http.StatusOK, session)
}

type diffRequest struct {
	FromSnapshotID *uuid.UUID `json:"from_snapshot_id"`
	ToSnapshotID   *uuid.UUID `json:"to_snapshot_id"`
}

func (h *BootstrapHandler) DiffSnapshots(c echo.Context) error {
	projectID, err := uuid.Parse(c.Param("projectId"))
	if err != nil {
		return ErrorResponse(c, http.StatusBadRequest, "INVALID_ID", "invalid project id", err.Error())
	}

	var req diffRequest
	if err := c.Bind(&req); err != nil {
		return ErrorResponse(c, http.StatusBadRequest, "INVALID_REQUEST", "failed to parse diff request", err.Error())
	}

	diff, err := h.service.DiffSnapshots(c.Request().Context(), projectID, req.FromSnapshotID, req.ToSnapshotID)
	if err != nil {
		return ErrorResponse(c, http.StatusInternalServerError, "INTERNAL_ERROR", "failed to diff snapshots", err.Error())
	}

	return SuccessResponse(c, http.StatusOK, diff)
}
